### ISSUE TITLE
Add Extenders to v1alpha2 Component Config

### ISSUE THREAD

**Issue by alculquicondor** [kubernetes/kubernetes#88529]
As `AlgorithmSource` is going away from the scheduler component config #87999, there will be no way for users to specify Extenders in this version.

Now that we have scheduling profiles #88087, we have some options:

1. Add `Extenders []Extender` to `KubeSchedulerConfiguration`. This implies that all profiles will share extenders. Let's recall that extenders are linked to custom resources. Requests are send to the extender if a pod has any resources declared as managed by the extender.
2. Add `Extenders []Extender` to `KubeSchedulerProfile`. Each profile can have its own.
3. Leave extenders unsupported in v1alpha2 for 1.18. Users can still use v1alpha1 in 1.18.

If we decide 1, we could later move to 2, but we have to introduce a new version (`v1alpha3` or `v1beta1`?). If we do 2, we shouldn't need to move back to 1, as users can copy and paste their configurations for each profile. But if that's what all users are doing, we might have to reconsider 1 for next release.
3 gives us one full release to gather feedback on the topic. It might even be the case that users just prefer to move to their extenders to the framework.

/sig scheduling

**Comment by ahg-g** [comment]
I discussed this with Aldo offline, my opinion is that we should go with option 1. Extenders already have an api to express whether or not it is interested in a pod, so we don't really need to have them in each profile. Also, moving from option 1 to 2 is easy in the future even after v1 since conversion is simple.

>  gives us one full release to gather feedback on the topic. It might even be the case that users just prefer to move to their extenders to the framework.
Framework and extender serve slightly different use cases, so I think they will coexist long term.

**Comment by alculquicondor** [comment]
cc @Huang-Wei

**Comment by alculquicondor** [comment]
From the SIG meeting, we shared the following thoughts on the options:

1. The downside is that users can't directly make some extenders optional from profiles. However, pods will preserve the ability to opt-out from an extender by not using their managed resources . In other words, this is no worst than 1.17 scheduler, which has no profiles support. Also, it is easier to move to option 2 in a future api version.
1. Gives too much burden to users who want to share extenders for all profiles. If we later learn that no one is setting different extenders per profile, it's hard to move back to option 1.
1. It's not a good idea to leave existing features unsupported from the new API.

Also @Huang-Wei proposed option 1.5:

Extenders are defined at the root, but profiles can simply specify which extenders they want to run by using a name key. The down side is that this requires more extensive api-review, for which we might not have enough time for 1.18. On the other hand, migrating from option 1 to this doesn't require a new API version, just new fields.

Additional notes from me:

1. The tricky case would be if an extender decides to manage CPU or memory. But this rather sounds like a bad practice.

If there are no additional concerns, I'll assign this.

**Comment by Huang-Wei** [comment]
Thanks @alculquicondor for the neat summarization.

BTW: let's also try to get feedback from end-users - I think @yqwang-ms, @everpeace and @cheyang are using extenders in their prod env. Can you please share your options?

**Comment by yqwang-ms** [comment]
Thanks @Huang-Wei  for the notification :)

Take the Motivation of Multi Scheduling Profiles:
> To serve such heterogeneous types of workloads...
> Instead, having a single kube-scheduler run multiple profiles will have the same benefits of running multiple schedulers without running into race conditions.

I see there are still gaps between multiple profiles and multiple schedulers, (I just know multiple profiles today, pls correct me if I am wrong)
1. If we using option 1, different profiles will share the same extender for the same interest resource name (such as abc.com/gpu). However, for multiple schedulers, different schedulers can config to use different extenders for the same interest resource name. 
For example, for the same abc.com/gpu, we have 2 OSS extenders OR 1 OSS extender but which can only load one conf profile for one process instance (similar to kubescheduler before multiple profiles is supported). One extender (or the profile) is for batch workload, another is for serving workload. To use and run them at the same time, one have to use multiple schedulers.
**So, overall, option 1 will also push extender's developer also need to support similar feature as "multiple profiles"**, so his extender does not have to run mutiple instances to serve mutiple workloads, so no need multiple schedulers anymore.
For our [HiveD scheduler extender](https://github.com/microsoft/pai/tree/master/subprojects/GOPATH/src/github.com/microsoft/hivedscheduler), we have already supported "multiple profiles", so not impacted by this.

2. Seems multiple profiles still using a single queue, for users (like us) which want **multiple queues** to avoid cross tenant starvation, still have to use multiple schedulers (and pointer to the same extender instance). (See more at https://github.com/kubernetes/kubernetes/issues/86373).
Anyway, seems all options here do not help to solve this issue.

Overall, both option 1 and 2 do not benefit or harm us, so I have no strong preference on this.
BTW, one idea is that, as option 1 is common case, only need to config once at one place, we should anyway support it even if option 2 is later supported at the same time (such as option 2 can override option 1's global config for a specific profile). So maybe we can start on option 1 now :), option 2 needs gather more feedbacks.

**Comment by alculquicondor** [comment]
It sounds like option 1.5 would work for you best. But as I said, that can be introduced in the next release without many complications.

Let's wait for others to comment. My only concern with 1.5 or 2 is that we are too close to code freeze.

Re 2:

Changing to a one-queue-per-profile model is something I thought about but it would have changed the architecture too much. Note that, because of that, we even restrict the use of a single queuesort plugin across profiles. There is room for improvement on that on next releases.

But even if we supported a queue per profile, it sounds like it wouldn't be enough for you, because you want to tie queues to tenants (which can be way more dynamic, whereas a profile would likely be just service and batch). So we probably need a solution tangential to profiles.

**Comment by yqwang-ms** [comment]
@alculquicondor, agree, I understand current situation, and all options are fine for me :)
For 2, we do not have to be one-queue-per-profile to fix the original issue #86373, and it is another topic which has less relation to current issue. Anyway, thanks for thinking about our requirements :)

**Comment by Huang-Wei** [comment]
Thanks @yqwang-ms for the feedback.

Along with multi-sched-profiles, the intuitive **N:1** situation comes to **N:M** (N = objects of such as extenders, and M = number of profiles), and we have to make reasonable decisions on the mapping. As @alculquicondor mentioned, I think the best bit for your case is to define top-level extenders, and then let each profile to choose which extender(s) to use. Due to the time limit, it's only possible to consider it in next releases.

Re 2: before thinking about mutli-queue-one-profile or one-queue-one-profile, I think we need to get a reasonable Queue API defined first, along with considering several factors:

- whether/how it should be tied to Pod or other workload API
- whether/how it should be associated with profiles (maybe they shouldn't be associated, as without a special QueueSort customization, a profile just handles how to schedule a Pod; while queue aims to control how/when to pop a Pod to let it be scheduled)
- whether/how it relates with tenants
- how it impacts current internal activeQ/backoffQ design
- how to pop out Pods from multiple queues
- ...

So basically multi-queue will inevitably introduce another dimension to current scheduler archi, we should follow the Kubernetes principle - no hidden API and make the Queue API reasonable, descriptive, unambiguous, only with that, it can benefit both default-scheduler out-of-tree scheduler plugins.

**Comment by alculquicondor** [comment]
@everpeace @cheyang any thoughts?

**Comment by alculquicondor** [comment]
/assign @damemi 
We need to do this ASAP

**Comment by fengzixu** [comment]
@damemi If you need my help, let me know. I am willing to do some work on this issue.

**Comment by damemi** [comment]
@alculquicondor working on it, I was just waiting to see if there was any more input

**Comment by alculquicondor** [comment]
Yeah, I was awaiting for more input too. But now we are way too close to code freeze.

According to the previous discussions, we should proceed with option 1 in 1.18, because:

- It doesn't require extensive api-review
- It is no worst than the existing functionality in 1.17
- It makes it easy to upgrade to other options in the next releases, once we have more input

**Comment by everpeace** [comment]
@alculquicondor thanks for mentioning me.  I agree option 1 for 1.18 because of schedule.  But I think option 1.5 would be natural for me.  So I think we should add extender field in the near future.
